[ROCm] upstream using rocprofiler-sdk (v3) for tracing AMD GPU events

cj401-amd opened the pull request:
@xla-rotation, would you please kindly help review this PR?

We are phasing out development and support for roctracer/rocprofiler/rocprof/rocprofv2 in favor of rocprofiler-sdk (v3) in upcoming ROCm releases. rocprofielr-sdk (v3) also moves away from cupti. 

This PR integrates rocprofiler-sdk (v3) into XLA for profiling GPU events on AMD GPUs. 

-  Integrate rocprofiler-sdk (v3) to XLA for improved profiling of GPU events, support both time-based and step-based profiling,
- Add unit tests for rocm_collector and rocm_tracer

cj401-amd commented on xla/backends/profiler/gpu/rocm_tracer.cc:
```diff
@@ -1588,3 +1606,599 @@ absl::Status RocmTracer::DisableActivityTracing() {
+}  // namespace profiler
+}  // namespace xla
+
+void __attribute__((constructor)) init_rocm_lib() {
+  rocprofiler_force_configure(xla::profiler::rocprofiler_configure);
+}
```
for some reason, we have to use this way to initialize rocprofiler-sdk. Properly we are going to fix this soon.

cj401-amd commented on xla/backends/profiler/gpu/rocm_tracer.h:
```diff
@@ -213,4 +213,115 @@ class RocmTracer {
+namespace xla {
+namespace profiler {
+
+std::string demangle(const char* name);
+std::string demangle(const std::string& name);
```
tsl::port::MaybeAbiDemangle works.

cj401-amd commented on xla/backends/profiler/gpu/rocm_tracer.h:
```diff
@@ -213,4 +213,115 @@ class RocmTracer {
+// The class use to enable rocprofiler-sdk buffered callback/activity tracing
+// and forward the collected trace events to RocmTraceCollector. There should be
+// only one RocmTracer per process.
+class RocmTracer {
```
we removed v1 for this PR.

cj401-amd committed:
update with refactoring based on the PR comments

cj401-amd commented on xla/backends/profiler/gpu/BUILD:
```diff
@@ -427,11 +441,62 @@ cc_library(
+    ],
+)
+
+xla_test(
```
after a refactoring to address dep issues. now it works as `xla_cc_test`.

cj401-amd committed:
update with VLOG(2) and fix a bug, new line at the EoF

cj401-amd committed:
update the macro

dimitar-asenov reviewed (changes requested):
Please `#include` in each source file only the files that are used in this file and all the files that are used in this file. E.g. `rocm_tracer_utils.cc` file at the moment includes only the corresponding header, but instead should include everything it uses directly. Similarly, `rocm_tracer_utils.h` should not include any files that it does not use directly. Please also do this for all the other files.

Another comment:
Do not use the `<chrono>` library. Use substitutes from `absl` instead.

cj401-amd committed:
update header files

dimitar-asenov commented on xla/backends/profiler/gpu/rocm_tracer.cc:
```diff
@@ -13,1578 +13,564 @@ See the License for the specific language governing permissions and
+#include <vector>
+#include <time.h>
+#include <unistd.h>
+#include <chrono>
```
Please do not use `<chrono>`. Use alternatives from `absl` instead.

dimitar-asenov commented on xla/backends/profiler/gpu/rocm_tracer.cc:
```diff
@@ -13,1578 +13,564 @@ See the License for the specific language governing permissions and
+#include <time.h>
+#include <unistd.h>
+#include <chrono>
+#include <unistd.h>  // For standard sysconf
```
This is already included a few lines above.

dimitar-asenov commented on xla/backends/profiler/gpu/rocm_tracer.cc:
```diff
@@ -13,1578 +13,564 @@ See the License for the specific language governing permissions and
+#include "xla/backends/profiler/gpu/rocm_tracer_utils.h"
 
 #include <cstdint>
+#include <cstring>
```
Please keep the entries within the individual include blocks sorted alphabetically.

dimitar-asenov commented on xla/backends/profiler/gpu/rocm_tracer.cc:
```diff
@@ -13,1578 +13,564 @@ See the License for the specific language governing permissions and
+    case ROCPROFILER_HIP_RUNTIME_API_ID_hipMemcpyWithStream:
+      return true;
+      break;
+    default:;
```
Please use `{}` to more explicitly denote an empty statement, as opposed to just `;`.

dimitar-asenov commented on xla/backends/profiler/gpu/rocm_tracer_utils.cc:
```diff
@@ -0,0 +1,102 @@
+    OO(Memset)
+    OO(Synchronization)
+    OO(Generic)
+    default:;
```
Please use `{}` to more explicitly denote an empty statement, as opposed to just `;`.

cj401-amd committed:
update a: addressing comments, b: add kernel details, c: update rocprofiler-sdk path

cj401-amd commented:
> > > Please `#include` in each source file only the files that are used in this file and all the files that are used in this file. E.g. `rocm_tracer_utils.cc` file at the moment includes only the corresponding header, but instead should include everything it uses directly. Similarly, `rocm_tracer_utils.h` should not include any files that it does not use directly. Please also do this for all the other files.
> > > Another comment: Do not use the `<chrono>` library. Use substitutes from `absl` instead.
> > 
> > 
> > Hey @dimitar-asenov, please help have a look as we have addressed your comments now.
> > btw, I wonder if it is possible to let me know what "Google internal checks FAILED" is?
> 
> Thank you for make the changes. I've flagged additional changes from our internal checks. Sorry for the back and forth, I hope this is the last round.

Hi @dimitar-asenov, I wonder if it is possible to let me know the internal check for this PR, so I can adapt the corresponding parts. Many thanks.

dimitar-asenov commented:
I tried updating our internal RoCM version but ran into some issues. I'm afraid we have to wait until Henning comes back next week. Sorry for the delay.

cj401-amd committed:
add rocprofiler-sdk debs for hermetic build

cj401-amd committed:
update with a successful hermetic build locally

cj401-amd commented:
> I tried updating our internal RoCM version but ran into some issues. I'm afraid we have to wait until Henning comes back next week. Sorry for the delay.

Hi @dimitar-asenov, thanks for looking after this. 

we did a hermetic build locally, which seems fine. 

`bazel build //xla/backends/profiler/gpu:rocm_tracer --config=rocm_ci_hermetic`

```
INFO: Analyzed target //xla/backends/profiler/gpu:rocm_tracer (191 packages loaded, 20450 targets configured).
[2,494 / 3,466] checking cached actions
[2,537 / 3,466] Compiling absl/strings/cord.cc; 0s local ... (127 actions, 126 running)
[2,667 / 3,466] Compiling absl/strings/cord.cc; 1s local ... (127 actions running)
[2,710 / 3,466] Compiling absl/strings/cord.cc; 2s local ... (128 actions running)
[2,754 / 3,466] Compiling src/google/protobuf/compiler/cpp/message_layout_helper.cc [for tool]; 3s local ... (128 actions, 127 running)
[2,779 / 3,466] Compiling src/google/protobuf/compiler/cpp/message_layout_helper.cc [for tool]; 4s local ... (128 actions, 127 running)
[2,788 / 3,466] Compiling src/google/protobuf/descriptor.cc; 4s local ... (129 actions, 128 running)
[2,802 / 3,466] Compiling src/google/protobuf/descriptor.cc; 6s local ... (128 actions, 127 running)
[2,833 / 3,466] Compiling src/google/protobuf/descriptor.cc; 7s local ... (128 actions, 127 running)
[2,891 / 3,466] Compiling src/google/protobuf/descriptor.cc; 8s local ... (129 actions, 127 running)
[2,941 / 3,466] Compiling src/google/protobuf/descriptor.cc; 9s local ... (125 actions, 124 running)
[3,064 / 3,466] Compiling src/google/protobuf/descriptor.cc; 10s local ... (126 actions, 124 running)
[3,154 / 3,466] Compiling src/google/protobuf/descriptor.cc; 11s local ... (127 actions, 126 running)
[3,310 / 3,466] Compiling src/google/protobuf/descriptor.cc; 12s local ... (30 actions running)
[3,333 / 3,466] Compiling src/google/protobuf/descriptor.cc; 13s local ... (7 actions running)
[3,337 / 3,466] Compiling src/google/protobuf/descriptor.cc; 14s local ... (3 actions running)
[3,338 / 3,466] Compiling src/google/protobuf/descriptor.cc; 16s local ... (2 actions running)
[3,339 / 3,466] Compiling src/google/protobuf/descriptor.cc [for tool]; 12s local
[3,340 / 3,466] [Prepa] Linking external/com_google_protobuf/upb_generator/minitable/protoc-gen-upb_minitable_stage0 [for tool]
[3,369 / 3,466] Compiling upb_generator/minitable/generator.cc [for tool]; 1s local ... (3 actions running)
[3,372 / 3,466] Linking external/com_google_protobuf/protoc [for tool]; 0s local
[3,406 / 3,466] Compiling src/google/protobuf/util/field_mask_util.cc; 0s local ... (57 actions running)
[3,437 / 3,466] Compiling src/google/protobuf/json/internal/untyped_message.cc; 2s local ... (26 actions running)
[3,465 / 3,466] Compiling xla/stream_executor/rocm/rocm_executor.cc; 4s local
INFO: Found 1 target...
Target //xla/backends/profiler/gpu:rocm_tracer up-to-date:
  bazel-bin/xla/backends/profiler/gpu/librocm_tracer.a
  bazel-bin/xla/backends/profiler/gpu/librocm_tracer.pic.a
  bazel-bin/xla/backends/profiler/gpu/librocm_tracer.so
INFO: Elapsed time: 139.259s, Critical Path: 23.84s
INFO: 3466 processes: 2500 internal, 966 local.
INFO: Build completed successfully, 3466 total actions
```

i-chaochen commented:
@cj401-amd you can rebase your branch to the latest xla.

cj401-amd committed:
update for fixing the typo

cj401-amd committed:
fix for hermetic build

alekstheod commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -317,7 +316,6 @@ def _find_libs(repository_ctx, rocm_config, bash_bin):
     Returns:
       Map of library names to structs of filename and path
     """
-    repo_path = str(repository_ctx.path(rocm_config.rocm_toolkit_path))
```
This merge broke our hermetic build
